### ISSUE TITLE
Issues with CJK Input

### ISSUE THREAD

**Issue by bitigchi** [skyjake/lagrange#417]
Text boxes do not react well to CJK input.

macOS 11.6.1, Lagrange 1.9.1

1. Japanese input is not processed at all, only IME suggestion box appears at an unrelated screen corner (top right), but the words do not display/not entered even if a suggestion is selected.
2. Chinese input is not displayed during selection phase, suggestion box appears on the bottom left corner, but once a selection is chosen it is displayed on the text box.

**Comment by skyjake** [comment]
This is related to #326.

**Comment by skyjake** [comment]
Not sure how to address this. It falls basically on SDL to handle the specifics of the IME in a platform-independent manner. I don't have any CJK proficiency so it's difficult for me to figure out how to get SDL to do the right thing here. I'll mark this as "help wanted" if someone wants to try to investigate/patch it a bit.

**Comment by cammeresi** [comment]
Regarding the K in CJK, when typing Korean, the input text box does not display the syllable under construction and only displays the completed syllable when typing has advanced such that the syllable boundary is unambiguous.  This makes typing somewhat difficult.

(Korean is written using an alphabet, and each syllable's letters are arranged into a block.  When typing in letters, the IME should rearrange the letters in the current syllable into the correct layout after each letter is typed.  E.g. if typing in the syllable 괜, the user types in ㄱㅗㅐㄴ and progressively sees ㄱ 고 괘 괜.)

Also, my keyboard shortcut to switch languages is Cmd-Option, but the application is, I guess, snooping on the modifier keys and sometimes sees the Option press as a cue to display link shortcuts, and it sometimes grabs input focus back from the input text box.

(Just FYI—I know you're not going to fix this soon.)

**Comment by fu-sen** [comment]
I have been able to confirm the same problem with Windows 10 + Japanese (Google Japanese Input). Lagrange 1.18.5
I'm using a text editor.

(In addition, I also tried Lagrange on Android + Gboard, and Japanese input works fine)

Oh, I'm a `balloon` in `bbs.geminispace.org`. 🎈

**Comment by tacomanator** [comment]
Hi @skyjake. New to Gemini. I only stumbled upon it when searching for Lagrange, the scientist and found your app in the process. I'm enjoying both Gemini and Lagrange so far, so thank you for that. Ran into this issue myself, and would like to offer advice.

Don't know about the IME suggestion box position and frankly it could be sticky, but is a higher priority change yet simple change you could make to open up the possibility of starting to actual use/test/and maybe resolve the other issues and that is to add an `Line Break ⇧↩︎  Accept ⌘↩︎` option to the `Return key behavior` setting.

In other words, an option which does nothing on bare return key presses.

The return key is essential to operating the IME selection box. CJK users need to be able to press enter without the app reacting. [Web]apps that listen to return key presses and submit, etc. are the bane of the asian computerist.

Is this something you'd be willing to add? I'd be happy to test and give feedback.

<img width="609" height="438" alt="Image" src="https://github.com/user-attachments/assets/c1e114de-2759-4e7d-b6ae-53cddfbcd090" />

**Comment by skyjake** [comment]
@tacomanator Adding that option is (was) straightforward. It will be included in the upcoming version 1.20.

**Comment by tacomanator** [comment]
@skyjake thank you so much!
I will try it out as soon as it lands.

**Comment by cammeresi** [comment]
If anyone is listening who speaks Chinese or Japanese, if you are able to build the latest Lagrange with a version of SDL modified with [my SDL patch](https://github.com/libsdl-org/SDL/pull/15251) and if you find the results positive, you may want to provide feedback there with the results of your testing, as they want to hear results about languages other than Korean before accepting the patch.

@fu-sen @tacomanator @bitigchi

**Comment by skyjake** [comment]
In v1.20.3, the "only with mods" return key behavior (mentioned above) will now affect the editor in UploadWidget as well.

**Comment by fu-sen** [comment]
![Image](https://github.com/user-attachments/assets/c32faa42-b47f-4d70-9a7d-1a879a4592bb)
I'm also getting the expected behavior for typing in Japanese!
There was no need to set the return (enter) key.

Related: https://github.com/skyjake/lagrange/pull/758

**Comment by aoiglobo** [comment]
At the request of @fu-sen, I tried a few Linux environments.

- When I using `fcitx5` and `fcitx` (old version) as the input method, it works fine.\
(Many CJK users probably use this)
- When I used Fedora GNOME and `ibus`, I was able to input Japanese as well.
(iBus is GNOME's default input method)
- I used Manjaro KDE Plasma, I couldn't input Japanese even though I was using `fcitx5` .\
I thought it was an effect of Wayland, but I used Raspberry Pi OS (LXDE) and it works fine. It wasn't.

Since Fcitx 5 and iBus are the current standards, many Linux CJK users will find this problem improving.

**Comment by tacomanator** [comment]
Tried via 1.20.3 on MacOS for Japanese input. Huge improvement, thank you!

I captured an issue in which, as text reaches the right side of the input, newly inputted characters are not visible if they are in an intermediate state. Notice some flashes to the right, as the IME briefly gives word suggestions.

https://github.com/user-attachments/assets/9c4fb9c8-a49b-4aeb-af65-38290da8bfae

In Japanese (and I assume Chinese/Korean), text is in an "intermediate" state, to allow for word selection, until "finalized." I assume calculation for the field size does not take this into account. As a side note, word selection is done by pressing enter, which is why input boxes that auto-submit on keypress=enter are hard to use.

In general I believe intermediate text entry would be accessible to the program. Example:

https://github.com/user-attachments/assets/add39871-ae9d-4d1e-bee4-ce60b2935f59

Text in intermediate state is underlined in blue.

If the intermediate text entry is not available for the size calculation, it would at least be better if the characters were visible even if they spill out of the right side of the dialog.

**Comment by cammeresi** [comment]
If you accept the patch I posted, I suggest that this bug be closed and kept closed lest it become a forum for any complaint about any IME-related behavior.  I suggest that opening new bugs for new complaints would be better.

**Comment by tacomanator** [comment]
I concur original issues—IME suggestion box placement and showing intermediate state text during entry—have been addressed. While somewhat related, the issue of the input field size during intermediate state text entry could be made into a new issue if you would prefer, just LMK.

**Comment by skyjake** [comment]
Ok, let's close this. Specific improvements can be opened as new issues.
